// cucumber-tsflow/dist/step-binding-flags.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // tslint:disable:no-bitwise
    /**
     * The CucumberJS step binding types.
     */
    var StepBindingFlags;
    (function (StepBindingFlags) {
        /**
         * No bindings.
         */
        StepBindingFlags[StepBindingFlags["none"] = 0] = "none";
        /**
         * A 'Given' step definition binding.
         */
        StepBindingFlags[StepBindingFlags["given"] = 1] = "given";
        /**
         * A 'When' step definition binding.
         */
        StepBindingFlags[StepBindingFlags["when"] = 2] = "when";
        /**
         * A 'Then' step definition binding.
         */
        StepBindingFlags[StepBindingFlags["then"] = 4] = "then";
        /**
         * A 'Before' hook binding.
         */
        StepBindingFlags[StepBindingFlags["before"] = 8] = "before";
        /**
         * An 'After' hook binding.
         */
        StepBindingFlags[StepBindingFlags["after"] = 16] = "after";
        /**
         * All step definition bindings.
         */
        StepBindingFlags[StepBindingFlags["StepDefinitions"] = 7] = "StepDefinitions";
        /**
         * All hook bindings.
         */
        StepBindingFlags[StepBindingFlags["Hooks"] = 24] = "Hooks";
    })(StepBindingFlags = exports.StepBindingFlags || (exports.StepBindingFlags = {}));
});
